src/cockpit/polkit.py

```diff
@@ -100,56 +100,52 @@ async def _communicate(self, process: asyncio.subprocess.Process):
-    After that, pass it to @unregister_agent().
+    Use this as a context manager to ensure that the agent gets unregistered again.
     """
-    try:
-        system_bus = bus.Bus.default_system()
-    except OSError as e:
-        logger.debug('cannot connect to system bus, not registering polkit agent: %s', e)
-        return None
-
-    try:
-        # may refine that with a D-Bus call to logind
-        subject = ('unix-session', {'session-id': {'t': 's', 'v': os.environ['XDG_SESSION_ID']}})
-    except KeyError:
-        logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
-        return None
-
-    agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(responder)
-    agent_slot = system_bus.add_object('/', agent_object)
-
-    # register agent
-    locale_name = locale.setlocale(locale.LC_MESSAGES, '')
-    await system_bus.call_method_async(
-        'org.freedesktop.PolicyKit1',
-        '/org/freedesktop/PolicyKit1/Authority',
-        'org.freedesktop.PolicyKit1.Authority',
-        'RegisterAuthenticationAgent',
-        '(sa{sv})ss',
-        subject, locale_name, '/')
-    logger.debug('Registered agent for %r and locale %s', subject, locale_name)
-    return {'subject': subject, 'slot': agent_slot}
-
-
-async def unregister_agent(agent):
-    """Unregister polkit agent.
-
-    You have to do this explicitly (not just rely on garbage collection), otherwise
-    the next register_agent() will fail with "there is already a registered agent".
-
-    This receives the agent object returned by @register_agent().
-    """
-    system_bus = bus.Bus.default_system()
-    await system_bus.call_method_async(
-        'org.freedesktop.PolicyKit1',
-        '/org/freedesktop/PolicyKit1/Authority',
-        'org.freedesktop.PolicyKit1.Authority',
-        'UnregisterAuthenticationAgent',
-        '(sa{sv})s',
-        agent['subject'], '/')
-    agent['slot'].cancel()
-    logger.debug('Unregistered agent for %r', agent['subject'])
+    def __init__(self, responder: InteractionResponder):
+        self.responder = responder
+        self.agent_slot = None
+
+    async def __aenter__(self):
+        try:
+            self.system_bus = bus.Bus.default_system()
+        except OSError as e:
+            logger.debug('cannot connect to system bus, not registering polkit agent: %s', e)
+            return self
+
+        try:
+            # may refine that with a D-Bus call to logind
+            self.subject = ('unix-session', {'session-id': {'t': 's', 'v': os.environ['XDG_SESSION_ID']}})
+        except KeyError:
+            logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
+            return self
+
+        agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(self.responder)
+        self.agent_slot = self.system_bus.add_object('/', agent_object)
+
+        # register agent
+        locale_name = locale.setlocale(locale.LC_MESSAGES, '')
```

Comment:
So this is tricky.  This is a statement with a heck of a side-effect, and it's going to happen somewhat randomly the first time we attempt to authenticate a user with polkit.  At this point in the code it's definitely inappropriate to call this function with anything other than `None` as the second argument.

I think we don't do `setlocale()` anywhere else, though, so there's a good chance you get `C` here.  It might make sense to call `setlocale(locale.LC_ALL, '')` somewhere early `main()`.  But this is fairly fraught — we still don't have a good story for language switching, unfortunately...

Reply:
Indeed that has bothered me for some time already: The bridge is supposed to run external programs under the user's locale, not 'C', to get translated output from them. The C bridge does that when loading packages by looking at the `Accept-Language:` header. Until we get that, I changed the second argument to None here -- at least this will start to DTRT as soon as we actually set the locale in the bridge in general.